### DISCUSSION TITLE
Make FunctionAuth part of the public API

### DISCUSSION THREAD

**Discussion by fsenart** [encode/httpx#1674] [Potential Issue]
It is surely an oversight, but the handy `auth.FunctionAuth` class is not exported.

**Comment by lovelydinosaur** [comment]
It was a deliberate choice, since you don't need to use the class for the simple function case, you can just pass the function itself.

Eg...

```python
def custom_auth(request):
    request["X-Auth-Token"] = "1234"

httpx.get("https://www.example.com", auth=custom_auth)
```

We *could* nonetheless expose the class, but as things currently stand it's an implementation detail, that users don't need to know about.

**Reply by zyv** [reply]
Just for the record:

```python
def custom_auth(request):
    request.headers["X-Auth-Token"] = "1234"
    return request
```

I came across this discussion via Google and was confused as to why the code wasn't working.

I have submitted a pull request to add this tip to the documentation:

https://github.com/encode/httpx/pull/3518

P.S. I got tripped up myself, so I think this needs to be added to the docs all the more.

**Comment by tscheburaschka** [comment]
Just for the record: This class is quite useful when implementing a sequential combination of auths as prepared in [`httpx_auth.authentication.SupportMultiAuth`](https://github.com/Colin-b/httpx_auth/blob/53e063c338ead7ba8a3442c7fe4efcb4befa057b/httpx_auth/authentication.py#L93). But it's fair to assume that users so deep in the rabbit hole are able to handle this case themselves.
